lib/signaling/v2/peerconnection.js

```diff
@@ -356,6 +380,31 @@ class PeerConnectionV2 extends StateMachine {
     this.emit('candidates', peerConnectionState);
   }
 
+  /**
+   * Handle an ICE connection state change event.
+   * @private
+   * @returns {void}
+   */
+  _handleIceConnectionStateChange() {
+    const { iceConnectionState } = this._peerConnection;
+
+    this._log.debug(`ICE connection state is "${iceConnectionState}"`);
+
+    // Case 1: Transition to "failed".
+    if (this._lastIceConnectionState !== 'failed' && iceConnectionState === 'failed' && !this._shouldRestartIce && !this._isRestartingIce) {
+      this._log.warn('ICE failed; attempting to restart ICE');
+      this._shouldRestartIce = true;
+      this.offer();
+    }
+
+    // Case 2: Transition from "failed".
+    else if (this._lastIceConnectionState === 'failed' && (iceConnectionState === 'connected' || iceConnectionState === 'completed')) {
+      this._log.info('ICE reconnected');
```

Comment:
@markandrus Should `._shouldRestartIce` again be set to `false` here?

Reply:
I'm not sure. The way I've implemented here, `iceConnectionState` transitioning to "failed" kicks off an ICE restart, and the ICE restart will be completed* regardless of whether `iceConnectionState` ever transitions out of "failed". I suppose if we were forced to do a rollback and, during that time, `iceConnectionState` transitioned out of "failed", we might be able to skip restarting ICE again... but I think this optimization would complicate the implementation. So I'd prefer not to set it here.

\* An ICE restart is completed once an offer produced by `createOffer({ iceRestart: true })` is answered, taking into account glare may force us to do a rollback before this happens.